### PR TITLE
[FLINK-4248] [core] [table] CsvTableSource does not support reading SqlTimeTypeInfo types

### PR DESCRIPTION
**Pull request by twalthr**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

This PR implements parsers for SQL Date/Time/Timestamp and tests. It also tests if the new parsers can be used in the RowCsvInputFormat.

### CODE REVIEW

**Comment by gallenvara** [comment]
Hi @twalthr , the logical of parserFields method in Date/Time/TimeStamp (also in the lastest PR for BigDecimal/BigInteger) is same, is it better to refactor by creating super class for them to reduce the duplicate codes?

**Comment by twalthr** [comment]
@gallenvara I thought the same. But I'm not sure if it is worth it. You could save ~25 lines of code per class maybe and if we want to make the parsing more efficient (do not use the `valueOf` methods, but own implementation) the super class becomes obsolete anyway.

File: flink-core/src/test/java/org/apache/flink/types/parser/SqlDateParserTest.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+		return new Date[] {
+			Date.valueOf("1970-01-01"), Date.valueOf("1990-10-14"), Date.valueOf("2013-08-12"),
+			Date.valueOf("2040-05-12"), Date.valueOf("2040-05-12"), Date.valueOf("1970-01-01")
+		};
+	}
+
+	@Override
+	public String[] getInvalidTestValues() {
+		return new String[] {
+			" 2013-08-12", "2013-08-12 ", "2013-08--12", "13-08-12", "2013/08/12", " ", "\t",
```
Comment (fhueske) [inline flink-core/src/test/java/org/apache/flink/types/parser/SqlDateParserTest.java]: Add a value with an invalid date like `"2000-02-30"`?

File: flink-core/src/test/java/org/apache/flink/types/parser/SqlTimeParserTest.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		return new Time[] {
+			Time.valueOf("00:00:00"), Time.valueOf("02:42:25"), Time.valueOf("14:15:51"),
+			Time.valueOf("18:00:45"), Time.valueOf("23:59:58"), Time.valueOf("0:0:0")
+		};
+	}
+
+	@Override
+	public String[] getInvalidTestValues() {
+		return new String[] {
+			" 00:00:00", "00:00:00 ", "00:00::00", "00x00:00", "2013/08/12", " ", "\t"
```
Comment (fhueske) [inline flink-core/src/test/java/org/apache/flink/types/parser/SqlTimeParserTest.java]: Add an invalid time `"12:00:61"`?

File: flink-core/src/test/java/org/apache/flink/types/parser/SqlTimestampParserTest.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+
+import java.sql.Timestamp;
+
+public class SqlTimestampParserTest extends ParserTestBase<Timestamp> {
+
+	@Override
+	public String[] getValidTestValues() {
+		return new String[] {
+			"1970-01-01 00:00:00.000", "1990-10-14 02:42:25.123", "1990-10-14 02:42:25.123000001",
```
Comment (fhueske) [inline flink-core/src/test/java/org/apache/flink/types/parser/SqlTimestampParserTest.java]: Add a valid time without second fractions: `"2016-05-01 12:34:50"`

File: flink-core/src/test/java/org/apache/flink/types/parser/SqlTimestampParserTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+			Timestamp.valueOf("2013-08-12 14:15:59.478"), Timestamp.valueOf("2013-08-12 14:15:59.47"),
+			Timestamp.valueOf("0000-01-01 00:00:00.000")
+		};
+	}
+
+	@Override
+	public String[] getInvalidTestValues() {
+		return new String[] {
+			" 2013-08-12 14:15:59.479", "2013-08-12 14:15:59.479 ", "1970-01-01 00:00::00",
+			"00x00:00", "2013/08/12", "0000-01-01 00:00:00.f00", "2013-08-12 14:15:59.4788888888888888",
```
Comment (fhueske) [inline flink-core/src/test/java/org/apache/flink/types/parser/SqlTimestampParserTest.java]: What is invalid about `"2013-08-12 14:15:59.4788888888888888"`, too much precision?
Add an invalid timestamp `"2010-12-12 15:67:01.123"`?

File: flink-core/src/main/java/org/apache/flink/types/parser/SqlDateParser.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+	private static final Date DATE_INSTANCE = new Date(0L);
+
+	private Date result;
+
+	@Override
+	public int parseField(byte[] bytes, int startPos, int limit, byte[] delimiter, Date reusable) {
+		int i = startPos;
+
+		final int delimLimit = limit - delimiter.length + 1;
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/SqlDateParser.java]: Should we move the logic of creating a String without leading or tailing space to `FieldParser`? A couple of parsers use that (the ones in this PR, BigInteger, ...).

File: flink-core/src/test/java/org/apache/flink/types/parser/SqlDateParserTest.java (hunk 1 above)
Comment (twalthr) [inline flink-core/src/test/java/org/apache/flink/types/parser/SqlDateParserTest.java]: The Java SQL date parser supports this and converts this into "2000-03-01".

Comment (fhueske) [inline flink-core/src/test/java/org/apache/flink/types/parser/SqlDateParserTest.java]: What does it make out of `"2000-14-01"`?

File: flink-core/src/test/java/org/apache/flink/types/parser/SqlTimeParserTest.java (hunk 2 above)
Comment (twalthr) [inline flink-core/src/test/java/org/apache/flink/types/parser/SqlTimeParserTest.java]: This is also not invalid. The seconds are converted to minutes.

File: flink-core/src/test/java/org/apache/flink/types/parser/SqlTimestampParserTest.java (hunk 4 above)
Comment (twalthr) [inline flink-core/src/test/java/org/apache/flink/types/parser/SqlTimestampParserTest.java]: Yes, the maximum precision is 9 digits.
The timestamp you suggest is also valid.

File: flink-core/src/main/java/org/apache/flink/types/parser/SqlDateParser.java (hunk 5 above)
Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/types/parser/SqlDateParser.java]: I will try to generalize it a bit.

File: flink-core/src/test/java/org/apache/flink/types/parser/SqlDateParserTest.java (hunk 1 above)
Comment (twalthr) [inline flink-core/src/test/java/org/apache/flink/types/parser/SqlDateParserTest.java]: That results in a `IllegalArgumentException`.

**Comment by twalthr** [comment]
@fhueske thanks for reviewing my code. I addressed your feedback.

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java [hunk 6]
```diff
@@ -174,6 +174,62 @@ protected void setErrorState(ParseErrorState error) {
...
+		int i = 0;
+		final byte delByte = (byte) delimiter;
+
+		while (i < length && bytes[startPos + i] != delByte) {
+			i++;
+		}
+
+		if (i > 0 &&
+				(Character.isWhitespace(bytes[startPos]) || Character.isWhitespace(bytes[startPos + i - 1]))) {
+			throw new NumberFormatException("There is leading or trailing whitespace in the numeric field.");
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: Throw a `IllegalArgumentException`? Method might be used by parsers that do not parse numeric values.

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java [hunk 7]
```diff
@@ -174,6 +174,62 @@ protected void setErrorState(ParseErrorState error) {
...
+		return len;
+	}
+
+	/**
+	 * Returns a string with a numeric format (like XXXX-XX-XX). Throws an exception if the
+	 * string contains leading/trailing whitespaces or if the column is empty.
+	 *
+	 * @return the parsed string
+	 */
+	public static final String formattedString(byte[] bytes, int startPos, int length, char delimiter) {
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: make `protected`?

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java [hunk 8]
```diff
@@ -174,6 +174,62 @@ protected void setErrorState(ParseErrorState error) {
...
 		return this.errorState;
 	}
+
+	/**
+	 * Returns the end position of a string with a numeric format (like XXXX-XX-XX). Sets the error state if the
+	 * string contains leading/trailing whitespaces or if the column is empty.
+	 *
+	 * @return the end position of the string or -1 if an error occurred
+	 */
+	public final int formattedStringEndPos(byte[] bytes, int startPos, int limit, byte[] delimiter) {
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: make `protected`?

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java [hunk 9]
```diff
@@ -174,6 +174,62 @@ protected void setErrorState(ParseErrorState error) {
...
+				(Character.isWhitespace(bytes[startPos]) || Character.isWhitespace(bytes[(len - 1)]))) {
+			setErrorState(ParseErrorState.NUMERIC_VALUE_ILLEGAL_CHARACTER);
+			return -1;
+		}
+
+		return len;
+	}
+
+	/**
+	 * Returns a string with a numeric format (like XXXX-XX-XX). Throws an exception if the
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: Why is this method specific for numeric formats?

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java [hunk 10]
```diff
@@ -174,6 +174,62 @@ protected void setErrorState(ParseErrorState error) {
...
+	 * string contains leading/trailing whitespaces or if the column is empty.
+	 *
+	 * @return the parsed string
+	 */
+	public static final String formattedString(byte[] bytes, int startPos, int length, char delimiter) {
+		if (length <= 0) {
+			throw new NumberFormatException("Invalid input: Empty string");
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: Change to `IllegalArgumentException`?

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java (hunk 7 above)
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: What is formatted about the returned string? How about renaming the method to `nextFieldAsString()`?

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java [hunk 11]
```diff
@@ -174,6 +174,62 @@ protected void setErrorState(ParseErrorState error) {
 	public ParseErrorState getErrorState() {
 		return this.errorState;
 	}
+
+	/**
+	 * Returns the end position of a string with a numeric format (like XXXX-XX-XX). Sets the error state if the
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: Why numeric format?

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java (hunk 8 above)
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: rename to something like `findNextFieldEndPos()`?

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java (hunk 11 above)
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: Maybe reduce the scope of the method to only identify the end of the field and do the whitespace check in each parser separately?

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java (hunk 9 above)
Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: Why not? Only parsers with a numeric format will use this method.

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java (hunk 10 above)
Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: According to the Javadoc "parseField" should return a `NumberFormatException`.

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java (hunk 6 above)
Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: I think we could just rename the method again. Maybe `numericString()`?

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java (hunk 9 above)
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: It returns a plain string object and only checks if the string has a leading or tailing whitespace. There are no checks for numeric characters or similiar. I think the method should be named according to what it does and not who's the intended caller.

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java (hunk 10 above)
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: But this is not the `parseField` method. `parseField` could catch the exception and pass it on as a `NumberFormatException`.

File: flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java (hunk 6 above)
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java]: But it does not do anything related to numeric values or am I overlooking something?

**Comment by twalthr** [comment]
@fhueske I think we should keep the methods as they currently are. I renamed them to "nextNumericString"/"nextNumericStringEndPos" and added a explanation what a numeric string is. You are right they don't do much numeric at the moment but only the numeric classes will use it. Furthermore, the error states and exceptions are numeric. E.g. if we don't do the whitespace checking in `nextNumericString` we also need to return the position instead of the string.

**Comment by fhueske** [comment]
I still think that methods should be named after what they do (and there is nothing numeric about their behavior) and not in which context they are supposed to be called. The error messages are numeric due to the original context of the code which has been moved into separate methods. That context is no longer present in these methods. Anyway, I don't want to start a bikeshedding discussion about the names of internal utility methods and stop at this point ;-)

**Pull request closed without merging**